Simplify and improve pipeliner error handling

nussjustin opened the pull request:
As reported in #189, the current `pipeliner` implementation does not handle panics when marshalling/unmarshalling and will kill the whole program. Fix this by recovering any panics from within the `pipelinerPipeline.Run` method.

Since I am here already I also refactored the error handling for the pipeliner which was more complex and subtle than it should have been.

Updates #189

mediocregopher commented on pipeliner.go:
```diff
@@ -194,15 +192,32 @@ func (p *pipeliner) flush(reqs []CmdAction) []CmdAction {
+	p.resSet = true
+}
+
+func (p *pipelinerCmd) send() {
```
Just for clarity I think `sendRes` might be a better name. Also, are there any cases where `resSet == true` and `p.res == nil`? I can't think how that would happen with the given logic, but maybe I'm missing something. If not, maybe it'd be simpler to drop `resSet` and just check that `res == nil` before writing it.

nussjustin replied:
`p.res` will be `nil` whenever the was no error, which should be most cases. If we were to drop `resSet` we would need to keep track where the `pipelinerPipine` exits with an error, otherwise we may set an error on a `pipelinerCmd` that had no error before.

Maybe another name than `resSet` might be better.

nussjustin commented:
I think we should do both. Handling panics in `resp.Any` is more generally useful, but having it also in the `pipeliner` does not hurt (it's basically just a `defer` with an `if`).

nussjustin commented:
Updated the PR with a (IMO) cleaner error tracking.

I also played with using values instead of pointers for `pipelinerCmd`s and found this to be a) a bit easier to read and b) much simpler since this automtically avoids all race conditions with the `*pipelinerCmd`s since it only pools the channels. I pushed this to [another branch here](https://github.com/nussjustin/radix/commit/8372478da811aec5c7c7e87b132d35eb0cd4c07a). The commit is a bit large because I renamed a few things. The only negative I could see about this is that I couldn't use the existing `pipeline` for this anymore and instead hat to re implement it (2 methods).

mediocregopher commented:
Reading through both I kind of like the current way a bit better, mostly because it's not re-implementing `Keys()`, and what you have here is easy enough to follow imo. But I'm down to merge whichever you prefer, both lgtm :)

nussjustin commented:
Let's go with the branch from this PR then for now. The change is smaller and simpler.

mediocregopher commented:
Done!
